R/summary_functions.R

```diff
@@ -133,6 +135,81 @@ zero_coverage <- function(x, ...) {
   }
 }
 
+
+
+
+#' Checks whether s1 contains s2
+#' s1: a srcref from branch_srcrefs
+#' s2: a srcref from covered_exp
+srcref_contains <- function (s1, s2) {
+    fst_ln1 <- s1[[1L]]
+    lst_ln1 <- s1[[3L]]
+    fst_col1 <- s1[[5L]]
+    lst_col1 <- s1[[6L]]
+
+    fst_ln2 <- s2[[1L]]
+    lst_ln2 <- s2[[3L]]
+    fst_col2 <- s2[[5L]]
+    lst_col2 <- s2[[6L]]
+
+    if (fst_ln1 < fst_ln2 & lst_ln1 > lst_ln2) {
+        TRUE
+    } else if (fst_ln1 == fst_ln2 & lst_ln1 == lst_ln2) {
+        (fst_col1 <= fst_col2 & lst_col1 >= lst_col2)
+    } else if (fst_ln1 == fst_ln2 && lst_ln1 > lst_ln2) {
+        (fst_col1 <= fst_col2)
+    } else if (fst_ln1 < fst_ln2 & lst_ln1 == lst_ln2) {
+        (lst_col1 >= lst_col2)
+    } else {
+        FALSE
+    }
+}
+
+
+
+
+
+#' Tally branch coverage
+#'
+#' @inheritParams print.branch_coverage
+#' @return a `data.frame` of branch coverage 
+#' @exports
+tally_branch_coverage <- function (x) {
+    brs <- attr(x, "branches")
+
+    covered_exp <- Filter(function(x) x$value > 0, x)
+    class(covered_exp) <- "list"           ##hacky: remove it later
+
+    for(exp in covered_exp) {
+        for (i in seq_along(brs)) {
+            sameFile <- (getSrcFilename(exp$srcref) == getSrcFilename(brs[[i]]$srcref))
+            if (!isTRUE(brs[[i]]$value) & sameFile) {
+                covered_br <- srcref_contains(unclass(brs[[i]]$srcref), unclass(exp$srcref))
+                brs[[i]]$value <- covered_br
+            }
+        }
+    }
+
+    if(length(brs) != 0) { 
+        df_br <- brs
+        class(df_br) <- "coverage"
+        as.data.frame(df_br)
+    } else {
+        as.data.frame(brs)  ## no branch -> empty data.frame
+    }
+
+
+}
+
+
+## srcref : List[refs]
+## objlist : List[records]
+## filteredobjlist =filter ( objlist , x => srcref contains x)
+## def filter(l : List[A], p : A => Bool): List[A] = l match {
```

Comment:
This looks like Scala :-)

Reply:
Oh this is pseudo code William wrote. Sorry, deleted!